haskell: Fix developPackage ignoring extend

Aver1y opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Previously `developPackage` would ignore `extend`, because `extensible-self`
would still refer to the unextended package set even after an `extend`,
i.e.:

```
(haskellPackages.extend f).developPackage = haskellPackages.developPackage
```

This fixes this by using `extends` on `self.__unfix__` instead of using
`extensible-self`. Since `extensible-self` wasn't otherwise used it was
removed.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cdepillabout commented on pkgs/development/haskell-modules/make-package-set.nix:
```diff
@@ -235,12 +231,14 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
-              overrides))
-        .callCabal2nix name root {};
+      let
+        extendFix' = f: fixed: fix' (extends f fixed.__unfix__);
```
Would you be able to add a comment here on what `extendFix'` is doing?  It is not immediately obvious to me what it is doing, or why it would be needed?

Aver1y replied:
> We would like to call `extend` on `self`, but that is not available. However for fixed points that were taken using `fix'` we can get the same effect by calling `extends` on the `__unfix__` attribute and taking the fixed point again.

Would that be OK?

cdepillabout commented:
@Aver1y Thanks for trying to fix this up.

Two things:

- Do you have a short reproducible example of how `developPackage` fails without this change? (And how it is fixed with this change?)
- Could you send this PR to the `haskell-updates` branch (instead of `master`).

-------------------

cc @peti, @Ericson2314, and @matthewbauer to get some more eyeballs on this.

Aver1y commented:
Here is an example (obviously doesn't actually build, but get's the point across):

**default.nix**

```nix
{ pkgs ? import <nixpkgs> {} }:

(pkgs.haskellPackages.extend (self: super: {
  base-with-a-strange-name = self.base;
})).developPackage {
  root = ./.;
}
```

**example.cabal**

```cabal
name:                developPackageExample
version:             0.1.0.0

library developPackageExample
  build-depends:       base-with-a-strange-name
```

**Before patch**
```
$ nix-build
error: anonymous function at [...] called without required argument 'base-with-a-strange-name', at [...]
```

**After**
```
$ nix-build --arg pkgs "import ../nixpkgs {}"
 derivations will be built:
  /nix/store/apdqihqw89l1z2d2sxcr02q0wz6dfjkx-developPackageExample-0.1.0.0.drv
building '/nix/store/apdqihqw89l1z2d2sxcr02q0wz6dfjkx-developPackageExample-0.1.0.0.drv'...
[...]
```

Aver1y committed:
haskell: Fix developPackage to work with extend

Previously developPackage would ignore extends, because extensible-self
would still refer to the unextended package set even after an extend,
i.e.:

(haskellPackages.extend f).developPackage = haskellPackages.developPackage

This fixes this by using extend on self.__unfix__ instead of using
extensible-self. Since extensible-self wasn't otherwise used it was
removed.

Aver1y commented:
Noticed the same pattern pop up here:

**make-package-set.nix:85**
```nix
let newScope = mkScope (fix' (extends f scope.__unfix__));
```

Aver1y commented:
Maybe we could even add it to `fixed-points.nix`?

teto committed:
haskellPackages.c2hsc: apply patchto find gcc

...used in tests.

Fix is upstream but there has been no sign of release for the past 6
months, fixes https://github.com/NixOS/nixpkgs/issues/63764.

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.4-4-gee9b477 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/fbfc752cf1c573ca69744af876072f5f9c847a78.

Commit:
haskellPackages.ginger: unbreak

peti committed:
LTS Haskell 16.0

peti committed:
Drop obsolete overrides now that LTS 16.x has arrived.

alexarice committed:
Agda: Fix version 2.6.1

maralorn committed:
haskellPackages.neuron: Fix build

peti committed:
haskell-polysemy-plugin: drop obsolete overrides

peti committed:
haskell-stack: drop obsolete overrides

http-download versios 2.0.x is the default in LTS 16.x.

peti committed:
haskell-src-exts-simple: drop obsolete overrides

peti committed:
haskell-postgresql-syntax: drop obsolete overrides

sorki committed:
haskellPackages.nix-narinfo: unbreak

sorki committed:
haskellPackages.zre: unbreak

sorki committed:
haskellPackages: add me as maintainer for my packages

and nix-derivation which I care about.

peti committed:
hackage2nix: update the set of broken packages to fix evaluation

peti committed:
hackage2nix.yaml: un-break maintained packages

Ping @sorki to please fix gcodehs and ttn-client.

peti committed:
hackage2nix: gitit is still broken, I'm afraid

Related to https://github.com/NixOS/nixpkgs/pull/89795.

peti committed:
all-cabal-hashes: update to Hackage at 2020-06-12T00:09:36Z

peti committed:
hackage2nix: gcodehs and ttn-client can't be build because of broken dependencies

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.4-4-gc6118bc from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/3984ed61bf06f0cd1b440a6be09c0cd20717ab9e.

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.5 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/0c8837fa12992012b1b3c577808768e7c303a9a6.

peti committed:
Merge remote-tracking branch 'origin/master' into haskell-updates.

Aver1y commented:
Ping?
